rust/common-graph/src/graph.rs

```diff
@@ -0,0 +1,731 @@
+        V: Clone,
+        'ext: 'a,
+    {
+        GraphProcessor::new(self, input)
+    }
+
+    /// Iteratively process this graph.
+    ///
+    /// `func` is called for each node with its inputs
+    /// in order to write its outputs.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use common_graph::{GraphProcessorItem, Graph, Result};
+    ///
+    /// # fn main() -> Result<()> {
+    /// #[derive(Debug)]
+    /// enum Op {
+    ///   Square,
+    /// }
+    ///
+    /// let nodes = [
+    ///     (None, "Root", (vec!["in".into()], vec!["out".into()])),
+    ///     (Some(Op::Square), "Square", (vec!["in".into()], vec!["out".into()])),
+    /// ];
+    /// let edges = [
+    ///     ((0, "out".into()), (1, "in".into())),
+    ///     ((1, "out".into()), (0, "in".into())),
+    /// ];
+    ///
+    /// fn callback<'a>(
+    ///    item: &'a mut GraphProcessorItem<'a, Op, f32>,
+    /// ) -> ::std::result::Result<(), String> {
+    ///     let v: Vec<_> = item
+    ///         .inputs()
+    ///         .iter()
+    ///         .filter_map(|(_, value)| value.as_ref().cloned())
+    ///         .collect();
+    ///     let inner = item.node().inner().ok_or(String::from("Missing node"))?;
+    ///     let result = match inner {
+    ///         Op::Square => v[0] * v[0],
+    ///     };
+    ///     let out_port: &mut (&str, &mut Option<f32>) = item
+    ///         .outputs()
+    ///         .iter_mut()
+    ///         .find(|(key, _)| *key == "out")
+    ///         .unwrap();
+    ///    *out_port.1 = Some(result);
+    ///     Ok(())
+    /// };
+    ///
+    /// let graph = Graph::new(nodes, edges)?;
+    /// let result = graph.process([("out", 5f32)], callback)??;
+    /// let root_out = result[0].0.clone();
+    /// assert_eq!(root_out, vec![("in", Some(25f32))]);
+    ///
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn process<'a, 'ext, I, V, F, E>(
+        &'a self,
+        inputs: I,
+        mut func: F,
+    ) -> Result<::std::result::Result<Vec<PortStore<'a, V>>, E>>
+    where
+        F: for<'b> FnMut(&'b mut GraphProcessorItem<'b, T, V>) -> ::std::result::Result<(), E>,
+        I: IntoIterator<Item = (&'ext str, V)>,
+        V: Clone + 'a,
+        T: Debug + 'a,
+        'ext: 'a,
+    {
+        let mut processor = self.process_iter(inputs)?;
+        while let Some(mut item) = processor.try_next()? {
+            if let Err(e) = func(&mut item) {
+                return Ok(Err(e));
+            };
+        }
+        let Some(output) = processor.output() else {
+            return Err(CommonGraphError::Unexpected(
+                "Graph does not have output.".into(),
+            ))?;
+        };
+        Ok(Ok(output))
+    }
+
+    /*
+    /// Iteratively process this graph with async callback.
+    ///
+    /// `func` is called for each node with its inputs
+    /// in order to write its outputs.
+    pub async fn process_async<'a, 'ext, I, V, F, Fut, E>(
```

Comment:
WIP, challenging with some HRTB constraints, will likely remove and skip for now as async is still supported